source/include/azure_iot_hub_client.h

```diff
@@ -0,0 +1,411 @@
+    eAzureIoTStatusPreconditionFailed = 412,
+    eAzureIoTStatusRequestTooLarge = 413,
+    eAzureIoTStatusUnsupportedType = 415,
+    eAzureIoTStatusThrottled = 429,
+    eAzureIoTStatusClientClosed = 499,
+    eAzureIoTStatusServerError = 500,
+    eAzureIoTStatusBadGateway = 502,
+    eAzureIoTStatusServiceUnavailable = 503,
+    eAzureIoTStatusTimeout = 504,
+} AzureIoTHubMessageStatus_t;
+
+/*
+ *  Cloud Message STRUCTS
+ */
+typedef struct AzureIoTHubClientCloudMessageRequest
+{
+    const void * pvMessagePayload;           /**< The pointer to the message payload. */
+    uint32_t ulPayloadLength;                /**< The length of the message payload. */
+
+    AzureIoTMessageProperties_t xProperties; /**< The bag of properties received with the message. */
+} AzureIoTHubClientCloudMessageRequest_t;
+
+/*
+ *  Method STRUCTS
+ */
+typedef struct AzureIoTHubClientMethodRequest
+{
+    const void * pvMessagePayload; /**< The pointer to the message payload. */
+    uint32_t ulPayloadLength;      /**< The length of the message payload. */
+
+    const uint8_t * pucRequestID;  /**< The pointer to the request ID. */
+    int16_t usRequestIDLength;     /**< The length of the request ID. */
+
+    const uint8_t * pucMethodName; /**< The name of the method to invoke. */
+    uint16_t usMethodNameLength;   /**< The length of the method name. */
+} AzureIoTHubClientMethodRequest_t;
+
+/*
+ *  Twin STRUCTS
+ */
+typedef struct AzureIoTHubClientTwinResponse
+{
+    AzureIoTHubMessageType_t messageType;     /**< The type of message received. */
+
+    const void * pvMessagePayload;            /**< The pointer to the message payload. */
+    uint32_t ulPayloadLength;                 /**< The length of the message payload. */
+
+    uint32_t ulRequestID;                     /**< The request ID for the twin response. */
+
+    AzureIoTHubMessageStatus_t messageStatus; /**< The operation status. */
+
+    const uint8_t * pucVersion;               /**< The pointer to the twin document version. */
+    uint16_t usVersionLength;                 /**< The length of the twin document version. */
+} AzureIoTHubClientTwinResponse_t;
+
+/* Typedef for the cloud message callback */
+typedef void ( * AzureIoTHubClientCloudMessageCallback_t ) ( struct AzureIoTHubClientCloudMessageRequest * pxMessage,
+                                                             void * pvContext );
+
+/* Typedef for the method callback */
+typedef void ( * AzureIoTHubClientMethodCallback_t ) ( struct AzureIoTHubClientMethodRequest * pxMessage,
+                                                       void * pvContext );
+
+/* Typedef for the twin callback */
+typedef void ( * AzureIoTHubClientTwinCallback_t ) ( struct AzureIoTHubClientTwinResponse * pxMessage,
+                                                     void * pvContext );
+
+/* Receive context to be used internally to the processing of messages */
+typedef struct AzureIoTHubClientReceiveContext
```

Comment:
thoughts on having an `internal/` directory in the include directory and putting private structs/functions in there?

```
│   azure_iot.c
│   azure_iot_hub_client.c
│   azure_iot_provisioning_client.c
│   CMakeLists.txt
│
├───include
│   │   azure_iot.h
│   │   azure_iot_config_defaults.h
│   │   azure_iot_hub_client.h
│   │   azure_iot_provisioning_client.h
│   │
│   └───internal
│           internal.h
│
└───interface
        azure_iot_mqtt.h
        azure_iot_transport_interface.h
```

cc @hihigupt

Reply:
How about naming file with _prv? 

This way developer who includes our project does not need to understand our folder struct and simply include header and interface required.

Reply:
We could take the same approach as the embedded and only need them to include one directory. In this case, the private header would be included via
```c
#include "internal/azure_middleware_internal.h"
```

Reply:
I really like the "internal/*" model.  This also makes our story to customers easier because we say "you can use stuff in the public part of folder" and there's some promises around back-compat there.  If you go into internal/... land then that should really almost be self-documenting in folder name not to do that, and we can add extra warnings in headers themselves of course too.

Reply:
There's also the question of whether we have a catchall subdirectory like we have in embedded. Helps with installation directories. So all our headers would be included via
```c
#include "azure_middleware_freertos/azure_iot_hub_client.h`
```
or something like that. So source directory would have `source/include/azure_middleware_freertos/stuff.h`. That intermediary name could be something different.

Reply:
will be tackled in another PR